skimage/morphology/_util.py

```diff
@@ -51,35 +54,188 @@ def _validate_connectivity(image_dim, connectivity, offset):
+        or column-major (Fortran-style) order.
 
     Returns
     -------
-    offsets : ndarray
+    raveled_offsets : ndarray
         Linear offsets to a samples neighbors in the raveled image, sorted by
-        their Euclidean distance from the center.
+        their distance from the center.
+
+    Notes
+    -----
+    This function will return values even if `image_shape` contains a dimension
+    length that is smaller than `selem`.
 
     Examples
     --------
     >>> _offsets_to_raveled_neighbors((4, 5), np.ones((4, 3)), (1, 1))
     array([-5, -1,  1,  5, -6, -4,  4,  6, 10,  9, 11])
     """
-    structure = structure.copy()  # Don't modify original input
-    structure[tuple(center)] = 0  # Ignore the center; it's not a neighbor
-    connection_indices = np.transpose(np.nonzero(structure))
-    offsets = (np.ravel_multi_index(connection_indices.T, image_shape,
-                                    order=order) -
-               np.ravel_multi_index(center, image_shape, order=order))
-    squared_distances = np.sum((connection_indices - center) ** 2, axis=1)
-    return offsets[np.argsort(squared_distances)]
+    if not selem.ndim == len(image_shape) == len(center):
+        raise ValueError(
+            "number of dimensions in image_shape, selem and center "
+            "does not match"
+        )
+
+    selem_indices = np.array(np.nonzero(selem)).T
+    offsets = selem_indices - center
+
+    if order == 'F':
+        offsets = offsets[:, ::-1]
+        image_shape = image_shape[::-1]
+    elif not order == 'C':
+        raise ValueError("order was not 'C' or 'F'")
+
+    # Scale offsets in each dimension and sum
+    ravel_factors = image_shape[1:] + (1,)
+    ravel_factors = np.cumprod(ravel_factors[::-1])[::-1]
+    raveled_offsets = (offsets * ravel_factors).sum(axis=1)
+
+    # Sort by distance
+    distances = np.abs(offsets).sum(axis=1)
+    raveled_offsets = raveled_offsets[np.argsort(distances)]
+
+    # In case any dimension in image_shape is smaller than selem.shape
+    # duplicates might occur, remove them
+    if any(x < y for x, y in zip(image_shape, selem.shape)):
+        # np.unique reorders, which we don't want
+        _, indices = np.unique(raveled_offsets, return_index=True)
+        raveled_offsets = raveled_offsets[np.sort(indices)]
+
+    # Remove "offset to center"
+    raveled_offsets = raveled_offsets[1:]
+
+    return raveled_offsets
+
+
+def _resolve_neighborhood(selem, connectivity, ndim):
+    """Validate or create structuring element.
+
+    Depending on the values of `connectivity` and `selem` this function
+    either creates a new structuring element (`selem` is None) using
+    `connectivity` or validates the given structuring element (`selem` is not
+    None).
+
+    Parameters
+    ----------
+    selem : array-like or None
+        The structuring element to validate. See same argument in
+        `local_maxima`.
```

Comment:
If  we move this function to util, this statement must be further explained.

Reply:
Done as well.